fasm/__init__.py

```diff
@@ -46,122 +60,134 @@ class ValueFormat(enum.Enum):
-        else:
-            start = int(set_feature_model.feature_address.address1, 10)
-            end = None
-            address_width = 1
+    feature, p = tagged_string_from_bytes('f', data, i)
+    start, end, p = fasm_address_from_bytes(data, p)
+    value, value_format, p = fasm_value_from_bytes(data, p)
 
-    if set_feature_model.feature_value:
-        width, value, value_format = verilog_value_to_int(
-            set_feature_model.feature_value)
+    if not value:
+        value = 1
 
-        if width is not None:
-            assert width <= address_width
-
-        assert value < (2**address_width), (value, address_width)
+    if end and not start:
+        start = end
 
     return SetFasmFeature(
-        feature=set_feature_model.feature,
-        start=start,
-        end=end,
-        value=value,
-        value_format=value_format,
-    )
-
-
-def get_fasm_metamodel():
-    return textx.metamodel_from_file(
-        file_name=os.path.join(os.path.dirname(__file__), 'fasm.tx'),
-        skipws=False)
-
-
-def fasm_model_to_tuple(fasm_model):
-    """ Converts FasmFile model to list of FasmLine named tuples. """
-    if not fasm_model:
-        return
-
-    for fasm_line in fasm_model.lines:
-        set_feature = None
-        annotations = None
-        comment = None
-
-        if fasm_line.set_feature:
-            set_feature = set_feature_model_to_tuple(fasm_line.set_feature)
-
-        if fasm_line.annotations:
-            annotations = tuple(
-                Annotation(
-                    name=annotation.name,
-                    value=annotation.value if annotation.value else '')
-                for annotation in fasm_line.annotations.annotations)
-
-        if fasm_line.comment:
-            comment = fasm_line.comment.comment
-
-        yield FasmLine(
-            set_feature=set_feature,
-            annotations=annotations,
-            comment=comment,
-        )
-
+        feature = feature,
+        start = start,
+        end = end,
+        value = value,
+        value_format = value_format
+    ), i+length
+
+# Decode an annotation: x = "y"
+def fasm_annotation_from_bytes(data, i):
+    length, i = get_header('a', data, i)
+    if not length:
+        return None, i
+
+    name, p = tagged_string_from_bytes('.', data, i)
+    value, p = tagged_string_from_bytes('=', data, p)
+    return Annotation(
+        name = name,
+        value = value
+    ), i+length
+
+# Decode a set of annotations: { ... }
+def fasm_annotations_from_bytes(data, i):
+    length, i = get_header('{', data, i)
+    if not length:
+        return None, i
+
+    annotations = []
+    annotation, p = fasm_annotation_from_bytes(data, i)
+    while annotation:
+        annotations.append(annotation)
+        annotation, p = fasm_annotation_from_bytes(data, p)
+
+    return annotations, i+length
+
+# Decode an entire FASM line
+def fasm_line_from_bytes(data, i):
+    length, i = get_header('l', data, i)
```

Comment:
Under what circumstances would this fail, except if `i == len(data)`?  Shouldn't all data within the string be contained within a `l` header?

Reply:
The format is designed to allow streaming lines, so while everything under a line has a size prefix (or the size is known), the stream of lines is null terminated. Nonetheless, this follows the same pattern as every other header, where a header mismatch gives `None` as the length. This naturally supports lists and optional data.

Reply:
Sorry, my question maybe wasn't clear.   Given the underlying structure of the stream, all data is contained within a line header, yes?  So by definition, if there are no more line headers, than it should be `eof(stream)`, yes?  What am I missing?

Reply:
You are correct that `i` should equal `length` at the end; I just wasn't clear what the suggested change is. I think you are suggesting `assert i == length`.

Reply:
Ya, that's what I was aiming for.